libcheckisomd5.c

```diff
@@ -32,336 +32,111 @@
-        } else {
-            loc++;
-        }
-
-        if ((skipfnd & md5fnd & fragsumfnd & fragcntfnd) & supportedfnd)
-            break;
-    }
-
-    if (!(skipfnd & md5fnd))
-        return -1;
-
-    /* get isosize */
-    *isosize = (buf[SIZE_OFFSET]*0x1000000+buf[SIZE_OFFSET+1]*0x10000 +
-                buf[SIZE_OFFSET+2]*0x100 + buf[SIZE_OFFSET+3]) * 2048LL;
-
-    return offset;
 }
 
-/* mediasum is the sum encoded in media, computedsum is one we compute   */
-/* both strings must be pre-allocated at least 33 chars in length        */
-static int checkmd5sum(int isofd, char *mediasum, char *computedsum, checkCallback cb, void *cbdata) {
-    int nread;
-    int i, j;
-    int appdata_start_offset, appdata_end_offset;
-    int nattempt;
-    int skipsectors;
-    int supported;
-    int current_fragment = 0;
-    int previous_fragment = 0;
-    unsigned int bufsize = 32768;
-    unsigned char md5sum[16];
-    unsigned char fragmd5sum[16];
-    unsigned int len;
-    unsigned char *buf;
-    long long isosize, offset, pvd_offset, apoff;
-    char fragmentsums[FRAGMENT_SUM_LENGTH+1];
-    char thisfragsum[FRAGMENT_SUM_LENGTH+1];
-    long long fragmentcount = 0;
-    MD5_CTX md5ctx, fragmd5ctx;
-
-    if ((pvd_offset = parsepvd(isofd, mediasum, &skipsectors, &isosize, &supported, fragmentsums, &fragmentcount)) < 0)
+static enum isomd5sum_status checkmd5sum(int isofd, checkCallback cb, void *cbdata) {
+    struct volume_info *const info = parsepvd(isofd);
+    if (info == NULL)
         return ISOMD5SUM_CHECK_NOT_FOUND;
 
-    /*    printf("Mediasum = %s\n",mediasum); */
-
-    /* rewind, compute md5sum */
-    lseek(isofd, 0L, SEEK_SET);
-
-    MD5_Init(&md5ctx);
+    const off_t total_size = info->isosize - info->skipsectors * SECTOR_SIZE;
+    if (cb)
+        cb(cbdata, 0, total_size);
 
-    offset = 0;
-    apoff = pvd_offset + APPDATA_OFFSET;
+    /* Rewind, compute md5sum. */
+    lseek(isofd, 0LL, SEEK_SET);
 
-    buf = malloc(bufsize * sizeof(unsigned char));
-    if (cb)
-        cb(cbdata, 0, isosize - skipsectors*2048);
+    MD5_CTX hashctx;
+    MD5_Init(&hashctx);
 
-    while (offset < isosize - skipsectors*2048) {
-        nattempt = MIN(isosize - skipsectors*2048 - offset, bufsize);
+    const size_t buffer_size = NUM_SYSTEM_SECTORS * SECTOR_SIZE;
+    unsigned char *buffer;
+    buffer = memalign((size_t) getpagesize(), buffer_size * sizeof(*buffer));
 
-        /*      printf("%lld %lld %lld %d\n", offset, isosize, isosize-SKIPSECTORS*2048, nattempt); */
+    size_t previous_fragment = 0UL;
+    off_t offset = 0LL;
+    while (offset < total_size) {
+        const size_t nbyte = MIN((size_t)(total_size - offset), buffer_size);
 
-        nread = read(isofd, buf, nattempt);
-        if (nread <= 0)
+        ssize_t nread = read(isofd, buffer, nbyte);
+        if (nread <= 0L)
             break;
 
-        if (nread > nattempt) {
-            nread = nattempt;
-            lseek(isofd, offset+nread, SEEK_SET);
```

Comment:
But there're other places where `read` is used in this library. Should this check be added in all places then? Also I'm afraid that it'll just confuse other developers who'll look at it. So if it lands it definitely has to land in it's own commit with a lot of explanation.